// @node-red/wasm-red/templates/wasm-wasi-node.js

const WASI = require("wasi");
const debug = "{{{ debug }}}" || false;

function getWasi(){
  const wasi = new WASI.WASI({
    version: 'preview1',
    args: [ /* empty for now */ ],
    env: { /* empty for now */ },
    preopens: {

      // TODO This should come from the compiler
      // Eitherway, it should be a separatede folder per file
      "./": "./",
    },
  });

  return wasi
}

// Solves the issue of serializing inherited and circular props
function flatten(obj){
  
  obj.flow = obj._flow;
  // TODO add here other needed info
  return obj
}

function replacer(key, value) {
  // Turn into a hash map
  const cache = new Set();
  return function(key, value) {
    if (typeof value === 'object' && value !== null) {
      if (cache.has(value)) {
        // Circular reference found
        return `[Circular][${key}]`; // or you can return a string like "[Circular]"
      }
      cache.add(value);
    }
    return value;
  };
}


function get_context(RED, COMMAND_QUEUE, node, instancecb, self, world = ""){
  return {
    "env": {

      {{#CAN_OPERATE_COMMANDS}}
      "node_red_pop_size": function() {
        let peek = COMMAND_QUEUE.pop();
        let json = JSON.stringify(peek);
        let buffer = new TextEncoder().encode(json);
        if(debug)  console.log("Sending command size", buffer.byteLength, `[${world}]`);
        COMMAND_QUEUE.push(peek);
        return buffer.byteLength;
      }.bind(self),
      {{/CAN_OPERATE_COMMANDS}}


      {{#CAN_OPERATE_COMMANDS}}
      "node_red_pop": function(data, offset, length) {
        // console.log("QUEUE", COMMAND_QUEUE);
        let json = JSON.stringify(COMMAND_QUEUE.pop());
        let buffer = new TextEncoder().encode(json);
        let bytes = new Uint8Array(instancecb().exports.memory.buffer, data + offset, length);
        bytes.set(buffer);
        //if(debug)  console.log("Sending command", data, offset, length, `[${world}]`);
        return buffer.byteLength;
      }.bind(self),
      {{/CAN_OPERATE_COMMANDS}}


      {{#CAN_READ_MSG}}
      "node_red_msg_size": function() {
          // Return the size in bytes of the message encoded as a JSON string 
          let json = JSON.stringify(msg);
          let buffer = new TextEncoder().encode(json);
          if(debug)  console.log("Sending msg size", buffer.byteLength, `[${world}]`);
          return buffer.byteLength;
      }.bind(self),
      {{/CAN_READ_MSG}}


      {{#CAN_READ_NODE}}
      "node_red_node_size": function() {
        // Return the size in bytes of the message encoded as a JSON string 
        let json = JSON.stringify(flatten(node), replacer());
        let buffer = new TextEncoder().encode(json);
        if(debug) console.log("Sending node struct size", buffer.byteLength, `[${world}]`);
        return buffer.byteLength;
      }.bind(self),
      {{/CAN_READ_NODE}}

      {{#CAN_READ_MSG}}
      "node_red_msg": function(data, offset, length){
        let json = JSON.stringify(msg);
        let buffer = new TextEncoder().encode(json);
        let bytes = new Uint8Array(instancecb().exports.memory.buffer, data + offset, length);
        bytes.set(buffer);
        if(debug)  console.log("Sending msg", data, offset, length, `[${world}]`);
        return buffer.byteLength;
      }.bind(self),
      {{/CAN_READ_MSG}}

      {{#CAN_READ_NODE}}
      "node_red_node": function(data, offset, length){
        // TODO we need to see the consequences of this
        //node.flow = node._flow;
        let json = JSON.stringify(flatten(node), replacer());
        //console.log("host", json);
        let buffer = new TextEncoder().encode(json);
        let bytes = new Uint8Array(instancecb().exports.memory.buffer, data + offset, length);
        bytes.set(buffer);
        //if(debug) console.log("Sending node", data, offset, length, `[${world}]`);
        return buffer.byteLength;
      }.bind(self),
      {{/CAN_READ_NODE}}

      {{#CAN_READ_CONTEXT}}
      "node_red_context_size": function() {
        // Return the size in bytes of the message encoded as a JSON string 
        let json = JSON.stringify(this.context());
        let buffer = new TextEncoder().encode(json);
        if(debug) console.log("Sending node struct size", buffer.byteLength, `[${world}]`);
        return buffer.byteLength;
      }.bind(self),
      {{/CAN_READ_CONTEXT}}

      {{#CAN_READ_CONTEXT}}
      "node_red_context": function(data, offset, length){
        let json = JSON.stringify(this.context());
        let buffer = new TextEncoder().encode(json);
        let bytes = new Uint8Array(instancecb().exports.memory.buffer, data + offset, length);
        bytes.set(buffer);
        if(debug) console.log("Sending node", data, offset, length, `[${world}]`);
        return buffer.byteLength;
      }.bind(self),
      {{/CAN_READ_CONTEXT}}

      {{#CAN_SEND}}
      "node_red_send": function(data, offset, length){
        if(debug)console.log("Sending data", data, offset, length);
        let d =  instancecb().exports.memory.buffer.slice(data + offset, data + offset + length);
        let buffer = new Uint8Array(d);
        let encoded = new TextDecoder().decode(buffer);
        
        msg = JSON.parse(encoded);
        if(debug) console.log("Calling node send", data, offset, length, `[${world}]`);
        instancecb().send(msg)
      }.bind(self),
      {{/CAN_SEND}}


      {{#CAN_EMIT}}
      "node_emit": function(data, offset, length){
        if(debug)console.log("Emitting data", data, offset, length);
        let d =  instancecb().exports.memory.buffer.slice(data + offset, data + offset + length);
        let buffer = new Uint8Array(d);
        let encoded = new TextDecoder().decode(buffer);
        
        msg = JSON.parse(encoded);
        RED.events.emit(msg.event, msg.payload);
      }.bind(self),
      {{/CAN_EMIT}}


      {{#CAN_DONE}}
      "node_red_done": function(data, offset, length){
        //console.log("Done data", data, offset, length);
        let d =  instancecb().exports.memory.buffer.slice(data + offset, data + offset + length);
        let buffer = new Uint8Array(d);
        let encoded = new TextDecoder().decode(buffer);
        // The message is changed here
        if(encoded){
            msg = JSON.parse(encoded);
            if(debug) console.log("Calling node done", offset, length, `[${world}]`);
            instancecb().done(msg)
        }
        else 
        {
          if(debug) console.log("Calling node done", offset, length, `[${world}]`);
          instancecb().done()
        }
      }.bind(self),
      {{/CAN_DONE}}

      {{#CAN_WARN}}
      "node_red_warn": function(data, offset, length){
        //console.log("Done data", data, offset, length);
        let d =  instancecb().exports.memory.buffer.slice(data + offset, data + offset + length);
        let buffer = new Uint8Array(d);
        let encoded = new TextDecoder().decode(buffer);
        console.log("Warn", encoded);
        // The message is changed here
        if(encoded){
            msg = JSON.parse(encoded);
            if(debug) console.log("Calling node done", data, offset, length, `[${world}]`);
            if(instancecb().warn)
              instancecb().warn(msg)
        }
      }.bind(self),
      {{/CAN_WARN}}

      {{#CAN_ERROR}}
      "node_red_error": function(data, offset, length){
        //console.log("Done data", data, offset, length);
        let d =  instancecb().exports.memory.buffer.slice(data + offset, data + offset + length);
        let buffer = new Uint8Array(d);
        let encoded = new TextDecoder().decode(buffer);
        console.log("Error", encoded);
        // The message is changed here
        if(encoded){
            msg = JSON.parse(encoded);
            if(debug) console.log("Calling node done", data, offset, length, `[${world}]`);
            if(instancecb().warn)
              instancecb().error(msg)
        }
      }.bind(self),
      {{/CAN_ERROR}}

      {{#CAN_PUSH_RESULT}}
      "node_red_result": function(data, offset, length){
        if(debug) console.log("Setting global result", `[${world}]`);
        // Set a custom value here
        let d =  instance().exports.memory.buffer.slice(data + offset, data + offset + length);
        let buffer = new Uint8Array(d);
        let encoded = new TextDecoder().decode(buffer);
        data = JSON.parse(encoded);

        GLOBAL_RESULT.push(data);
      }.bind(self),
      {{/CAN_PUSH_RESULT}}

      "node_red_register": function(data, offset, length){
        // Set a custom value here
        let d =  instancecb().exports.memory.buffer.slice(data + offset, data + offset + length);
        let buffer = new Uint8Array(d);
        let encoded = new TextDecoder().decode(buffer);
        data = JSON.parse(encoded);

        
        if(debug) console.log("[Wrapper] Registering node", data, `[${world}]`);
        if(debug) console.log("[Wrapper] making a stateful instance", instancecb(), `[${world}]`);
        function wrappernode(newnode){
           RED.nodes.createNode(this, newnode);
           let augmentednode = this;
           // console.log(this._flow);
           //console.log(this, newnode, augmentednode);
            
           // Here we instantiate a node with the same properties as the one we are wrapping
           if(debug) console.log("Calling node constructor");
           let [wasm, address] = RED.GLOBAL_CACHE["{{{ wasmfile }}}"];
           // Should be save the instance to keep a state ?
           // TODO do we need a fresh instance of WASI ?
           const wasi = getWasi();
           // TODO do we need a fresh instance of Wasm
           const instance = new WebAssembly.Instance(wasm, {      
            'wasi_snapshot_preview1': wasi.wasiImport,
              ...get_context(RED, COMMAND_QUEUE, augmentednode, () => instance /**/, this, "instance")
            });
            
           COMMAND_QUEUE.push({
              "type": "init",
              "name": data.constructor
            })

            instance.warn = augmentednode.warn;
            wasi.start(instance);
            instance.exports.javyinit();


           // instance.exports.invoke to initialize the instance and its initial state
           // TODO replace by a wrapper function
           this.on('input', function(msg, send, done, warn) {
              
              instance.send = send;
              instance.done = done;
              instance.warn = warn || augmentednode.warn ;
              COMMAND_QUEUE.push({
                "type": "on",
                "event": "input",
                "msg": msg,
                "cl": data.constructor
              })
              
              instance.exports.javyprocessmessage();
              // instance.exports.on("input", msg);
           });
        };
        RED.nodes.registerType(data.name, wrappernode, data.options);
        if(debug) console.log("[Wrapper] Node registered", `[${world}]`);
      }.bind(self)

   }
  }
}

module.exports = function(RED){
    
    // Since the entry point of the module is always the same, we need to 
    // address their behavior by passing messages to it
    let COMMAND_QUEUE = [];
    
    if(debug) console.log("[Wrapper] Initializing context. Reading Global cache", RED.GLOBAL_CACHE);
    let [wasm, address] = RED.GLOBAL_CACHE["{{{ wasmfile }}}"];
    
    const wasi = getWasi();

    if(debug) console.log("[Wrapper] Wasi loaded");
    // The instantiation of the Wasm module can be avoided every time 
    // if for example, we have a pool of instances previously
    const instance = new WebAssembly.Instance(wasm, {
      'wasi_snapshot_preview1': wasi.wasiImport,
      // This is a static stage, so, no node instace neither Wasm instance
      ...get_context(RED, COMMAND_QUEUE, null, () => instance, this, "register")
    });
    
    if(debug) console.log("[Wrapper] Instance created. Calling start to invoke the node registration");
    // Start the timer here to se the impact of the instantiation
    // Move it along to include other computations, for example the call to WebAssembly instantiate
    COMMAND_QUEUE.push({
      "type": "register",
    })
    wasi.start(instance);
    console.log("[Wrapper] calling register node inside the Wasm file");
    instance.exports.javyregisternode()
}
