// @node-red/nodes/core/parsers/70-JSON.bck.js
/**
 * Copyright JS Foundation and other contributors, http://js.foundation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 **/

module.exports = function(RED) {
    "use strict";
    const Ajv = require('ajv');
    const ajv = new Ajv({allErrors: true});
    ajv.addMetaSchema(require('ajv/lib/refs/json-schema-draft-06.json'));

    console.log("This is testing");

    function JSONNode(n) {
        RED.nodes.createNode(this,n);
        this.indent = n.pretty ? 4 : 0;
        this.action = n.action||"";
        this.property = n.property||"payload";
        this.schema = null;
        this.compiledSchema = null;

        var node = this;
        console.log("Logging from inside the JSONNode function");

        this.on("input", function(msg,send,done) {
            
        });
    }
    RED.nodes.registerType("json",JSONNode);
}
